README

'mem' is a C++ implementation of the F77 code memsys.



ChangeLog

This is a fixed package and I don't expect any changes.
	
	

src/test.f
      program test
      double precision d(6,6),v(6),vec(6,6)
      d(1,1) = 1.;
      d(1,2) = 67788.2
      d(2,1) = d(1,2)
      d(1,3) = -0.024412
      d(3,1) = d(1,3)
      d(2,2) = 1.
      d(2,3) = 6628.52
      d(3,2) = d(2,3)
      d(3,3) = 1.
      call meml33(3,d,v,vec)
      print *, (v(i),i=1,3)
      end
      
      
      SUBROUTINE MEML33(ND,AMAT,VAL,VEC)
*+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
*     
*     Purpose:
*     Diagonalisation of symmetric matrix
*     k      k         k
*     Solves      AMAT(i,j) * VEC(j)  = VAL  * VEC(i)
*     for eigenvalues VAL and eigenvectors VEC, for k=1,2,...,ND
*     
*     Parameters:
*     ARGUMENT TYPE  I/O  DIMENSION   DESCRIPTION
*     ND      I     I      -         Number of dimensions
*     AMAT    R*8   I     ND,ND      Matrix (from 6*6 array)
*     VAL     R*8     O    ND        Eigenvalues
*     VEC     R*8     O   ND,ND      Eigenvectors
*     
*     Globals:
*     VARIABLE  COMMON BLOCK  I/O  DESCRIPTION
*     -
*     
*     External calls:
*     -
*     
*     History:
*     John Skilling    8 Nov 1985     Initial release
*     
*     
*     Notes:
*     (1) Eigenvalues VAL are returned in increasing order
*     (2) Eigenvectors VEC are normalised to V.AMAT.V=1
*     (3) Input matrix AMAT is preserved
*     (4) Only the upper triangle j>=i of AMAT(i,j) is read
*     
*     (5) Algorithm is to repeatedly square AMAT until the largest
*     eigenvalue dominates, then to subtract off that eigenvector,
*     and repeat until all the eigenvectors have been found.
*     This ND**4 algorithm would be too slow for large matrices,
*     but is robust for small.
*     
*-----------------------------------------------------------------------
      IMPLICIT CHARACTER*1 (A-H,O-Z)
      REAL*8 AMAT(6,6),VAL(6),VEC(6,6)
      REAL*8 Z(6,6),P(6,6),Q(6,6),X(6),Y(6),T,A,B,C,EPS
*     EPS is related to machine accuracy
      PARAMETER (EPS=3.0E-16)
*     Copy AMAT to a positive definite matrix Z (by adding C to eigenvals)
      A=0.0
      B=0.0
      T=0.0
      DO 2 I=1,ND
         A=A+AMAT(I,I)
         DO 1  J=1,ND
            C=AMAT(I,J)
            IF(ABS(C).LT.1.0E-18) C=0.0
            B=B+C**2
 1       CONTINUE
         T=T+1.0
    2 CONTINUE
      C=MAX(B-A*A/T,0.0D0)
      C=A/T-SQRT(C)-EPS*SQRT(B)
      print *,c
      DO 4 I=1,ND
         DO 3 J=I,ND
            Z(I,J)=AMAT(I,J)
            Z(J,I)=AMAT(I,J)
 3       CONTINUE
         Z(I,I)=Z(I,I)-C
    4 CONTINUE
* LMAX = maximum number of inner loop iterates
      T=-LOG(EPS)/(EPS*LOG(2.0))
      T=LOG(T)/LOG(2.0)
      LMAX=IFIX(SNGL(T+2.0))
*
      N=ND
* Outer loop over N for successively smaller eigenvalues
    5 CONTINUE
        T=0.0
        DO 7 I=1,ND
          DO 6 J=1,ND
            P(I,J)=Z(I,J)
    6     CONTINUE
          T=T+P(I,I)
    7   CONTINUE
        
        print *,"t = ",t
*
        L=0
* Inner loop over L for squaring P and setting T=trace
    8   CONTINUE
          L=L+1
          T=1.0/T
          DO 10 I=1,ND
            DO  9 J=1,ND
              Q(I,J)=P(I,J)*T
              IF(ABS(Q(I,J)).LT.1.0E-18) Q(I,J)=0.0
    9       CONTINUE
   10     CONTINUE
          T=0.0
          DO 13 I=1,ND
            DO 12 J=I,ND
              A=0.0
              DO 11 K=1,ND
                A=A+Q(I,K)*Q(K,J)
   11         CONTINUE
              IF(ABS(A).LT.1.0E-18) A=0.0
              P(I,J)=A
              P(J,I)=A
   12       CONTINUE
            T=T+P(I,I)
   13     CONTINUE
          print *,"end t = ",t
        IF( T.LT.1.0-EPS*10.0 .AND. L.LE.LMAX )  GOTO 8
* End inner loop when P is dyadic
*
* K = largest column = estimate of current largest eigenvector
        K=1
        A=0.0
        DO 14 I=1,ND
          IF(P(I,I).GT.A) THEN
            A=P(I,I)
            K=I
          ENDIF
   14   CONTINUE
*
* P(P(largest column)) = better estimate X of eigenvector
        DO 16 I=1,ND
          A=0.0
          DO 15  J=1,ND
            A=A+P(I,J)*P(J,K)
   15     CONTINUE
          IF(ABS(A).LT.1.0E-18) A=0.0
          Y(I)=A
   16   CONTINUE
        T=0.0
        DO 18 I=1,ND
          A=0.0
          DO 17 J=1,ND
            A=A+P(I,J)*Y(J)
   17     CONTINUE
          IF(ABS(A).LT.1.0E-18) A=0.0
          T=T+A*A
          X(I)=A
   18   CONTINUE
* Repeat..  Orthogonalise X to previous eigenvectors
        K=ND
        IF(K.EQ.N) GOTO 22
   19     A=0.0
          DO 20 I=1,ND
            A=A+X(I)*VEC(I,K)
   20     CONTINUE
          DO 21 I=1,ND
            B=X(I)-A
            IF(ABS(B).LT.1.0E-18) B=0.0
            X(I)=B
   21     CONTINUE
          K=K-1
          IF(K.GT.N) GOTO 19
* ..while
* Normalise eigenvector X
   22   A=0.0
        DO 23 I=1,ND
          A=A+X(I)*X(I)
   23   CONTINUE
        A=1.0/SQRT(A)
* Copy eigenvector X into output array VEC
        DO 24 I=1,ND
          X(I)=A*X(I)
          VEC(I,N)=X(I)
   24   CONTINUE
* Set eigenvalue VAL directly from the eigenvector
        A=0.0
        DO 26 I=1,ND
          DO 25 J=1,ND
            B=X(I)*X(J)
            IF(ABS(B).LT.1.0E-18) B=0.0
            A=A+Z(I,J)*B
   25     CONTINUE
   26   CONTINUE
        VAL(N)=A+C
* Finish ?   (if full set of eigenvectors has been found)
        N=N-1
        IF(N.LE.0) RETURN
* Otherwise, remove dyadic  X.Xtranspose  from matrix Z
        DO 28 I=1,ND
          DO 27 J=I,ND
            B=X(I)*X(J)
            IF(ABS(B).LT.1.0E-18) B=0.0
            Z(I,J)=Z(I,J)-A*B
            IF(ABS(Z(I,J)).LT.1.0E-18) Z(I,J)=0.0
            Z(J,I)=Z(I,J)
            print *, "z ", z(i,j)
   27     CONTINUE
   28   CONTINUE
      GOTO 5
* End outer loop
*
      END





src/test.cc

#include "../base/subs.h"
#include "../base/graphics.h"
#include "trm_memsys.h"

const int NDATA = 200;
float dat[NDATA];

void op(float model[], float data[] , const int npt);
void tr(float data[] , float model[], const int npt);

int main(){

  const int MXPIX = 100000;
  float buff[MXPIX];
  Mem::Gbl::st = buff;
  
  Mem::memcore(MXPIX,NDATA,NDATA);

  float model[NDATA], data[NDATA], x[NDATA], e[NDATA], image[NDATA];
  for(int i=0; i<NDATA; i++){
    x[i] = i+1;
    model[i] = 1.;
  }
  for(int i=0; i<NDATA; i++){
    model[i] += 2.*exp(-0.2*sqr(i-56.));
    model[i] += 4.*exp(-0.5*sqr(i-106.));
    model[i] += 4.*exp(-0.1*sqr(i-146.));
  }

  op(model,data,NDATA);
  long int seed = -8269;
  float sigma = 0.01;
  for(int i=0; i<NDATA; i++){
    data[i] += sigma*gauss2(seed);
    e[i]     = sigma;
    image[i] = 0.5;
  }

  // move stuff into vectors

  for(int i=0; i<NDATA; i++){
    Mem::Gbl::st[Mem::Gbl::kb[0]+i]  = image[i];
    Mem::Gbl::st[Mem::Gbl::kb[20]+i] = data[i];
    Mem::Gbl::st[Mem::Gbl::kb[21]+i] = 2./(e[i]*e[i]*NDATA);
  }
  
  float caim = 1., rmax = 0.2, c,test,acc,cnew,s,rnew,snew,sumf;
  for(int it=0; it<20; it++){
    std::cout << "\nIteration " << it+1 << std::endl;
    Mem::memprm(10,20,caim,rmax,1.,acc,c,test,cnew,s,rnew,snew,sumf);
  }

  // move stuff out for plot

  for(int i=0; i<NDATA; i++){
    image[i] = Mem::Gbl::st[Mem::Gbl::kb[0]+i];
  }

  try{
    graphics plot("/xs");
    plot.sci(4);
    plot.scf(2);
    plot.sch(1.5);
    plot.env(0.,float(NDATA+1),0.,6.,0,0);
    plot.sci(2);
    plot.lab("X","Y");
    plot.sci(3);
    plot.line(NDATA,x,model);
    plot.sci(2);
    plot.line(NDATA,x,image);
    plot.sci(1);
    plot.bin(NDATA,x,data,true);
    plot.clos();
  }

  catch(graphics::no_device o){
    std::cerr << "graphics error. No device open";
    std::string s = o.get_mess();
    if(s != ""){
      std::cerr << " in " << s << "." << std::endl;
    }else{
      std::cerr << "." << std::endl;
    }
  }

  catch(graphics::graphics_error o){
    std::cerr << "graphics error.";
    std::string s = o.get_mess();
    if(s != ""){
      std::cerr << s << std::endl;
    }else{
      std::cerr << std::endl;
    }
  }

}

void Mem::opus(const int j, const int k){

  std::cout << "    OPUS " << j+1 << " ---> " << k+1 << std::endl;
  op(Mem::Gbl::st+Mem::Gbl::kb[j],
     Mem::Gbl::st+Mem::Gbl::kb[k], NDATA);
}

const int NSP = 15;
float wgt[NSP] = {0.5,1.,1.5,2.,3.,4.,5.,6.,5.,4.,3.,2.,1.5,1.,0.5};

void op(float model[], float data[], const int npt){ 
  int k;
  float sum, wnorm = 0.;

  for(int i=0; i<NSP; i++) 
    wnorm += wgt[i];
  for(int i=0; i<npt; i++){
    sum = 0.;
    k = i - NSP/2;
    for(int j=0; j<NSP; j++, k++){
      if(k<0){
	sum += wgt[j]*model[0];
      }else if(k >= npt){
	sum += wgt[j]*model[npt-1];
      }else{
	sum += wgt[j]*model[k];
      }
    }
    data[i] = sum/wnorm;
  }
}


void tr(float data[], float model[], const int npt){

  int k;
  float wnorm = 0.;

  for(int i=0; i<NSP; i++)
    wnorm += wgt[i];

  for(int i=0; i<npt; i++)
    model[i] = 0.;

  for(int i=0; i<npt; i++){
    k = i - NSP/2;
    for(int j=0; j<NSP; j++, k++){
      if(k<0){
	model[i] += wgt[j]*data[0];
      }else if(k >= npt){
	model[i] += wgt[j]*data[npt-1];
      }else{
	model[i] += wgt[j]*data[k];
      }
    }
    model[i] /= wnorm;
  }
}




src/memcore.f
	SUBROUTINE MEMCORE(MXBUF, NXPIX, NYPIX, NZPIX, NOBS, IER)
C
C  FORMATS MEMSYS CORE BUFFER FOR IMAGES AND DATA VECTORS
C
C  Input:
C	MXBUF	= MAXIMUM ALLOWED NUMBER OF 4-BYTE SLOTS IN STORAGE BUFFER
C	NXPIX	= X DIMENSION OF IMAGES
C	NYPIX	= Y DIMESNION OF IMAGES
C	NZPIX	= Z DIMENSION OF IMAGES
C  Output:
C	IER	= 1 IF REQUIRED DIMENSIONS OVERFLOW MXBUF, 0 OTHERWISE.
C
C  FEB 1984 BY KDH AT IOA
C
	COMMON/MECOMP/NJ,MJ,NK,MK,KA(40),KB(40),KC(40),KD(40),
     *IOUT,L0,L1,M0,M1,M2,M3,M4
C
	INTEGER*4 JFILE(20),KFILE(20)

	DATA JFILE/
     * 1, 2, 3, 4, 5, 6, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 9, 7, 8/


	DATA KFILE/
     * 1, 2, 3, 4, 5, 6, 7, 8, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0/
	DATA MXJFILE,MXKFILE/20,20/
C
	WRITE(*,*) ' '
	WRITE(*,*) 'FORMATTING CORE STORAGE FOR DATA AND IMAGE VECTORS.'
	WRITE(*,*) 'IMAGE DIMENSION (X,Y,Z)=',NXPIX,NYPIX,NZPIX
	WRITE(*,*) ' DATA DIMENSION =',NOBS
C
C  LOAD IMAGE POINTERS
C
   25	NJ=1
	MJ=NXPIX*NYPIX*NZPIX
	NJFILE=0
	KORE=0
      DO I=1,MXJFILE
	KB(I)=-9999999
      IF(JFILE(I).GT.0) THEN
	NJFILE=NJFILE+1
	KB(I)=(JFILE(I)-1)*NJ*MJ + 1
	KORE=MAX(KORE,KB(I))
      END IF
      END DO
	KOREJ=KORE + NJ*MJ - 1
C
C  LOAD DATA POINTERS
C
	NK=1
	MK=NOBS
	NKFILE=0
      DO I=1,MXKFILE
	KB(MXJFILE+I)=-9999999
      IF(KFILE(I).GT.0) THEN
	NKFILE=NKFILE+1
	KB(MXJFILE+I)=KOREJ + (KFILE(I)-1)*NK*MK+1
	KORE=MAX(KORE,KB(MXJFILE+I))
      END IF
      END DO
C
C  SUMMARIZE STORAGE REQUIREMENTS ON TERMINAL
C
	KORE = KORE + NK*MK - 1
	KOREK= KORE - KOREJ
	NIMAGES=KOREJ/(NJ*MJ)
	NDATAS =KOREK/(NK*MK)
	WRITE(*,*) NJFILE,' IMAGES IN',NIMAGES,' FILES'
	WRITE(*,*) (KB(I),I=1,MXJFILE)
	WRITE(*,*) NKFILE,' DATA VECTORS IN',NDATAS,' FILES'
	WRITE(*,*) (KB(I),I=MXJFILE+1,MXJFILE+MXKFILE)
	NTEST = KB(NJFILE+MXKFILE) + NK*MK
	WRITE(*,*) 'IMAGE STORE:',KOREJ,' =',FLOAT(KOREJ)/KORE*100.,' %'
	WRITE(*,*) ' DATA STORE:',KOREK,' =',FLOAT(KOREK)/KORE*100.,' %'
	WRITE(*,*) 'TOTAL STORE:',KORE,' =',FLOAT(KORE)/MXBUF*100.,' %'
     *	,' OF BUFFER:',MXBUF
C
C  BUFFER OVERFLOWS
C
      IF(KORE .GT. MXBUF) THEN
	WRITE(*,*) CHAR(7),
     &'STORAGE REQUIREMENTS OVERFLOW THE BUFFER:',KORE,MXBUF
	IER = 1
	RETURN
      END IF
C
C  NORMAL RETURN
C
	IER = 0
	RETURN
	END


src/memedit.f
	SUBROUTINE MEMEDIT( MODE, LEVEL, DEF, AIM, RMAX, ACC )
C
C  PROGRAM FOR INTERACTIVELY SELECTING MEM ITERATION PARAMETERS
C
C  JULY 83 BY KDH AT IOA
* Dec 85 KDH at IOA - revised for new release of MEMSYS

	CHARACTER*1 PARAM

    1	TYPE *,' '
	TYPE *,'MEM iteration parameters:'
	TYPE *,'        MODE',MODE
	TYPE *,'       LEVEL',LEVEL
	TYPE *,'         AIM',AIM
	TYPE *,'        RMAX',RMAX
	TYPE *,'     DEFAULT',DEF
	TYPE *,' '

* decode mode

	M4=MODE/10000
	M3=(MODE-10000*M4)/1000
	M2=(MODE-10000*M4-1000*M3)/100
	M1=(MODE-10000*M4-1000*M3-100*M2)/10
	M0=(MODE-10000*M4-1000*M3-100*M2-10*M1)

* report search directions

    5 IF(M0.EQ.0) THEN
	TYPE *,'      full recursive generation of search directions'
      ELSE IF(M0.EQ.3) THEN
	TYPE *,'      3 search directions (plus current map)'
      ELSE IF(M0.EQ.4) THEN
	TYPE *,'      4 search directions (plus current map)'
      ELSE IF(M0.EQ.6) THEN
	TYPE *,'      6 search directions (plus current map)'
      ELSE
	TYPE *,CHAR(7),'INVALID NUMBER OF SEARCH DIRECTIONS', M0
	M0 = 0
	GOTO 5
      END IF

* report type of default

   10 IF(M1.EQ.0) THEN
	TYPE *,'      default =', DEF
      ELSE IF(M1.EQ.1) THEN
	TYPE *,'      default = exp(total(plnf))'
      ELSE IF(M1.EQ.3) THEN
	TYPE *,'      default = user image'
      ELSE IF(M1.EQ.4) THEN
	TYPE *,'      default =', DEF, ' with 0<f<1'
      ELSE IF(M1.EQ.5) THEN
	TYPE *,'      default =',DEF,' with total(f) held fixed'
      ELSE IF(M1.EQ.6) THEN
	TYPE *,'      default = exp(total(plnf)) with total(f) held fixed'
      ELSE IF(M1.EQ.8) THEN
	TYPE *,'      default = user image with total(f) held fixed'
      ELSE
	TYPE *,CHAR(7),'INVALID DEFAULT OPTION', M1
	M1 = 1
	GOTO 10
      END IF

* report constraint option 

   15 IF(M2.EQ.0) THEN
	TYPE *,'      chi-squared =', AIM
      ELSE IF(M2.EQ.1) THEN
	TYPE *,'      chi-squared = ', AIM, ' with sigma =', ACC
      ELSE IF(M2.EQ.5) THEN
	TYPE *,'      chi-squared rejecting outliers = ', AIM
      ELSE IF(M2.EQ.6) THEN
	TYPE *,'      chi-squared rejecting outliers = ', AIM, ' sigma =', ACC
      ELSE
	TYPE *,CHAR(7),'** INVALID CONSTRAINT OPTION', M2
	M2 = 1
	GOTO 15
      END IF

* report weight option

   20 IF(M3.EQ.0) THEN
	TYPE *,'      equal pixel weights'
      ELSE IF(M3.EQ.1) THEN
	TYPE *,'      user supplied pixel weights'
      ELSE
	TYPE *,CHAR(7),'** INVALID WEIGHT OPTION', M3
	M3 = 0
	GOTO 20
      END IF

	MODE = 1000*M3 + 100*M2 + 10*M1 + M0

* command node

	TYPE *,' '
	TYPE '(A)', '$What parameter to change (<CR> to proceed) ? '
	READ(*,'(A)') PARAM

	IF( PARAM .EQ. ' ' ) RETURN

      IF( PARAM .EQ. 'M' ) THEN
	TYPE *,'old MODE : ', MODE, ' enter new MODE : '
	READ(*,*,ERR=777) NEW
	MODE=NEW

      ELSE IF( PARAM .EQ. 'L' ) THEN
	TYPE *,'old LEVEL : ', LEVEL,' enter new LEVEL : '
	READ(*,*,ERR=777) NEW
	LEVEL=NEW

      ELSE IF( PARAM .EQ. 'D' ) THEN
	TYPE *,'old DEFAULT : ', DEF,' enter new DEFAULT : '
	READ(*,*,ERR=777) VALUE
	IF(VALUE.LE.0.) GOTO 777
	DEF=VALUE

      ELSE IF( PARAM .EQ. 'A' ) THEN
	TYPE *, 'old AIM : ', AIM, ' enter new AIM : '
	READ(*,*,ERR=777) VALUE
	IF(VALUE.LE.0.) GOTO 777
	AIM=VALUE

      ELSE IF( PARAM .EQ. 'R' ) THEN
	TYPE *,'old RMAX : ', RMAX, ' enter new RMAX : '
	READ(*,*,ERR=777) VALUE
	IF(VALUE.LE.0.) GOTO 777
	RMAX=VALUE

      END IF
	GOTO   1

  777	TYPE *, CHAR(7), '** INVALID REPLY, OLD VALUE KEPT.'
	GOTO   1
      END


src/testf.f
      PROGRAM TESTF
      
      INTEGER NDATA
      PARAMETER (NDATA = 80)
      REAL MODEL(NDATA), DATA(NDATA), X(NDATA)
      REAL E(NDATA), IMAGE(NDATA), GAUSS2
      INTEGER SEED/-8269/
      REAL  SIGMA/0.01/
      REAL CAIM, RMAX, C, TEST, CNEW, S, RNEW, ACC, SNEW, SUMF
      INTEGER MODE/10/
      INTEGER LEVEL/20/
      COMMON /MECOMP/ NJ,MJ,NK,MK,KA(40),KB(40),KC(40),KD(40),
     * IOUT,IIN,L0,L1,M0,M10,M11,M20,M21,M3
      COMMON /MECOMS/ ST(100000)
      DATA IOUT/6/

      CALL MEMCORE(100000,NDATA,1,1,NDATA,IER)
      IF(IER.NE.0) STOP 'FAILED TO FORMAT CORE'

      DO I=1,NDATA
         X(I) = I+1
         MODEL(I) = 1.
      END DO
      
      MODEL(40) = 3.
      MODEL(41) = 3.
      MODEL(39) = 2.
      MODEL(42) = 2.
      MODEL(38) = 1.5
      MODEL(43) = 1.5
  
      CALL OP(MODEL,DATA,NDATA)

      DO I = 1, NDATA
         DATA(I)  = GAUSS2(DATA(I),SIGMA,SEED)
         E(I)     = SIGMA
         IMAGE(I) = 0.5
      END DO

      DO I = 1, NDATA
         ST( KB(1)+I-1) = IMAGE(I)
         ST(KB(21)+I-1) = DATA(I)
         ST(KB(22)+I-1) = 2./(E(I)*E(I)*NDATA)
      END DO

      CAIM = 1.
      RMAX = 0.2
      
      DO I = 1, 10
         CALL MEMPRM( MODE, LEVEL, CAIM, RMAX, 1., ACC, C, TEST,
     &        CNEW, S, RNEW, SNEW, SUMF)
      END DO

      DO I = 1, NDATA
         IMAGE(I) = ST(KB(1)+I)
      END DO
      END

      SUBROUTINE OPUS(J, K)
      COMMON /MECOMP/ NJ,MJ,NK,MK,KA(40),KB(40),KC(40),KD(40),
     * IOUT,IIN,L0,L1,M0,M10,M11,M20,M21,M3
      COMMON /MECOMS/ ST(100000)
      INTEGER NDATA
      PARAMETER (NDATA = 80)

      WRITE(*,*) '     OPUS ',J,' ---> ',K
      CALL OP(ST(KB(J)),ST(KB(K)), NDATA)
      RETURN
      END


      SUBROUTINE OP(MODEL, DATA, NPT) 
      INTEGER NSP, NPT
      REAL DATA(NPT), MODEL(NPT)
      PARAMETER (NSP=9)
      REAL WGT(NSP)
      DATA WGT/1.,2.,4.,5.,6.,5.,4.,2.,1./
      INTEGER K
      REAL SUM, WNORM
      
      WNORM = 0
      DO I=1,NSP
         WNORM = WNORM + WGT(I)
      END DO
      DO I=1,NPT
         SUM = 0.
         K = I - NSP/2
         DO J = 1, NSP
            IF(K.LT.1) THEN
               SUM = SUM + WGT(J)*MODEL(1)
            ELSE IF(K .GT. NPT) THEN
               SUM = SUM + WGT(J)*MODEL(NPT)
            ELSE
               SUM = SUM + WGT(J)*MODEL(K);
            END IF
            K = K + 1
         END DO
         DATA(I) = SUM/WNORM
      END DO
      RETURN
      END

      SUBROUTINE TROPUS(K, J)
      COMMON /MECOMP/ NJ,MJ,NK,MK,KA(40),KB(40),KC(40),KD(40),
     * IOUT,IIN,L0,L1,M0,M10,M11,M20,M21,M3
      COMMON /MECOMS/ ST(100000)
      INTEGER NDATA
      PARAMETER (NDATA = 80)

      WRITE(*,*) '   TROPUS ',J,' <--- ',K
      CALL TR(ST(KB(K)),ST(KB(J)), NDATA)
      RETURN
      END


      SUBROUTINE TR(DATA, MODEL, NPT) 
      INTEGER NSP, NPT
      REAL DATA(NPT), MODEL(NPT)
      PARAMETER (NSP=9)
      REAL WGT(NSP)
      DATA WGT/1.,2.,4.,5.,6.,5.,4.,2.,1./
      INTEGER K
      REAL SUM, WNORM
      
      WNORM = 0
      DO I=1,NSP
         WNORM = WNORM + WGT(I)
      END DO
      DO I=1,NPT
         MODEL(I) = 0.
      END DO
      DO I=1,NPT
         K = I - NSP/2
         DO J = 1, NSP
            IF(K.LT.1) THEN
               MODEL(I) = MODEL(I) + WGT(J)*DATA(1)
            ELSE IF(K .GT. NPT) THEN
               MODEL(I) = MODEL(I) + WGT(J)*DATA(NPT)
            ELSE
               MODEL(I) = MODEL(I) + WGT(J)*DATA(K)
            END IF
            K = K + 1
         END DO
         MODEL(I) = MODEL(I)/WNORM
      END DO
      RETURN
      END


